### ISSUE TITLE
catch exceptions in boolyness of check

### ISSUE THREAD

**Issue by jamescasbon** [Parquery/icontract#105]
This is very much related to #97 

Consider this: 

```
$ cat i.py
from icontract import require, ensure
import numpy as np

@require(lambda x: np.array([True, False]))
def frobulate(x: int) -> int:
    return x

$ python -c "import i; i.frobulate(1)"                                                                                                                                                                                                 Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/james/.conda/envs/time_gym35/lib/python3.5/site-packages/icontract/_checkers.py", line 297, in wrapper
    _assert_precondition(contract=contract, resolved_kwargs=resolved_kwargs)
  File "/home/james/.conda/envs/time_gym35/lib/python3.5/site-packages/icontract/_checkers.py", line 91, in _assert_precondition
    if not check:
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

```

If you are doing numeric programming with icontract it's very likely at some point you will mess up and return a bool vector rather than a bool.  At the moment, its impossible to know where you did this.  If you have many requirements you have no decent way of debugging them.  You might end up commenting them out one by one or moving them to plain old asserts in the function.

Its annoying that numpy arrays are not true or false but as such a large part of the ecosystem, perhaps we could catch the value error during contract evaluation...

i.e.

    try: 
        falsiness = not check 
     except ValueError:
        print('useful info on where the exception occurred')
        raise from None

    if falsisness:

**Comment by mristin** [comment]
Hi @jamescasbon,
Thanks for noticing this! As we don't use numpy that intensively, we miss a lot of these use cases.

Actually, we can just try/except a general exception `err` when testing falsiness and then `raise ... from err`.

I'm on holidays till May 6th. Please feel free to submit a pull request or I'll fix this issue after the holidays.

**Comment by jamescasbon** [comment]
See #106

**Comment by mristin** [comment]
Hi @jamescasbon ,
I'll have a look at this issue tomorrow (2019-05-09 CET). I'll take the code from #106 and make a separate, more atomic pull request since some changes were merged in from the pull request #109 .

**Comment by mristin** [comment]
Hi @jamescasbon ,
I'm looking into the issue right now. Sorry for the delay.
